kafka-connect-gcp-storage/src/main/scala/io/lenses/streamreactor/connect/gcp/storage/storage/GCPStorageDirectoryLister.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2017-2024 Lenses.io Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.lenses.streamreactor.connect.gcp.storage.storage
+
+import cats.effect.IO
+import com.google.cloud.storage.Storage
+import com.google.cloud.storage.Storage.BlobListOption
+import com.typesafe.scalalogging.LazyLogging
+import io.lenses.streamreactor.connect.cloud.common.config.ConnectorTaskId
+import io.lenses.streamreactor.connect.cloud.common.model.location.CloudLocation
+import io.lenses.streamreactor.connect.cloud.common.storage.DirectoryLister
+
+import scala.jdk.CollectionConverters.IterableHasAsScala
+
+class GCPStorageDirectoryLister(connectorTaskId: ConnectorTaskId, storage: Storage)
+    extends LazyLogging
+    with DirectoryLister {
+
+  /**
+    * @param wildcardExcludes allows ignoring paths containing certain strings.  Mainly it is used to prevent us from reading anything inside the .indexes key prefix, as these should be ignored by the source.
+    */
+  override def findDirectories(
+    bucketAndPrefix:  CloudLocation,
+    recurseLevels:    Int,
+    exclude:          Set[String],
+    wildcardExcludes: Set[String],
+  ): IO[Set[String]] = {
+
+    def listSubdirs(prefix: String, recurseLevels: Int): Iterable[String] = {
```

Comment:
I think this method can be made tail-recursive

```scala
def listSubdirs(prefix: String, recurseLevels: Int): Iterable[String] = {

  @tailrec
  def loop(prefixes: List[String], acc: List[String]): Iterable[String] = {
    if (prefixes.isEmpty || recurseLevels <= 0) {
      acc
    } else {
      val foundResults = prefixes.flatMap { p =>
        val blobListOptions = BlobListOption.dedupe(
          BlobListOption.delimiter("/"),
          BlobListOption.pageSize(1000),
          BlobListOption.prefix(p),
          BlobListOption.currentDirectory(),
        )

        storage
          .get(bucketAndPrefix.bucket)
          .list(blobListOptions: _*)
          .iterateAll()
          .asScala
          .filter(_.isDirectory)
          .map(_.getName)
          .toList
          .filter { prefix =>
            connectorTaskId.ownsDir(prefix) && !exclude.contains(prefix) && !wildcardExcludes.exists(we =>
              prefix.contains(we),
            )
          }
      }

      val nextPrefixes = foundResults.filter(_ => recurseLevels > 1)

      loop(nextPrefixes, acc ++ foundResults)
    }
  }

  loop(List(prefix), List())
}
```

i haven't validated it, write on paper kind of approach.

Reply:
Love it, and wanted to use it, but couldn't get it working 😢